app/controllers/ems_common.rb

```diff
@@ -200,6 +200,15 @@ def ad_hoc_metrics_pressed
     javascript_redirect(polymorphic_path(@record, :display => "ad_hoc_metrics"))
   end
 
+  # provider_id can be either a provider id or a storage manager id, depending on context
+  # this method figures the storage_manager_id.
+  def storage_manager_id(provider_id)
+    manager = find_record_with_rbac(ExtManagementSystem, provider_id)
+    return nil unless manager
+    return manager.id unless manager.respond_to?(:storage_managers)
+    manager.storage_managers.detect { |m| m.supports_block_storage? }&.id
+  end
```

Comment:
I like this. Maybe rename function into <pre>def <b>block</b>_storage_manager_id</pre> because at some point we'll probably need <pre>def <b>object</b>_storage_manager_id</pre> as well. Or even better, make it accept additional parameter to support both storage manager types:

```ruby
def storage_manager_id(provider_id, criteria: :supports_block_storage?)
  ...
  manager.storage_managers.detect { |m| m.send(criteria) }&.id
end
```

I'll test your latest changes tomorrow when I come to the office.